Add basic documentation about shrink/split/clone APIs (#140139)
elastic/elasticsearch: docs/internal/DistributedArchitectureGuide.md

### Create a Shard

### Local Shards Recovery

> Local shards recovery is a type of recovery that reuses existing data from another shard(s) allocated on the current node (hence local shards). It is used exclusively to implement index [Shrink/Split/Clone APIs](#shrinksplitclone-index-apis).

Reviewer: ```suggestion
Local shards recovery is a type of recovery that reuses existing data from other shard(s) allocated on the current node (hence local shards). It is used exclusively to implement index [Shrink/Split/Clone APIs](#shrinksplitclone-index-apis).
```

### Closing a Shard

(this can also happen during shard reallocation, right? This might be a standalone topic, or need another section about it in allocation?...)

# Shrink/Split/Clone index APIs

> These APIs allow to create a new index that contains a copy of data from a provided index and differs in number of shards and/or index settings. They can only be executed on indices that are marked read-only.

Reviewer 2: ```suggestion
These APIs are used to create a new index that contains a copy of data from a provided index and differs in number of shards and/or index settings. They can only be executed against source indices that are marked read-only.
```

> Shrink API (`/{index}/_shrink/{target}`) allows to create an index that has fewer shards than the original index.

Reviewer 2: ```suggestion
The shrink API (`/{index}/_shrink/{target}`) creates an index that has fewer shards than the original index.
```

Split API (`/{index}/_split/{target}`) allows to create an index that has more shards than the original index.

> Clone API (`/{index}/_clone/{target}`) allows to create an index that has different set of index settings but the same number of shards as the original index.

Reviewer 2: ```suggestion
The clone API (`/{index}/_clone/{target}`) creates an index that has the same number of shards as the original index but may have different index settings.
```

The main implementation logic is centralized in `TransportResizeAction` however it only creates new index in the cluster state using special `recoverFrom` and `resizeType` parameters. The entire workflow involves multiple components.

The high level structure is the following:

> 1. `TransportResizeAction` creates index metadata for the new index that contains information about the resize performed. It also creates a routing table for the new index which based on the resize information in the index metadata assigns `LocalShardsRecoverySource.INSTANCE` recovery source to primary shards.

Reviewer 2: ```suggestion
1. `TransportResizeAction` creates index metadata for the new index that contains information about the resize performed. It also creates a routing table for the new index which assigns a `LocalShardsRecoverySource.INSTANCE` recovery source to primary shards based on the resize information in the index metadata .
```
